### PR TITLE
Added command to rename current document, with lsp support

### PR DESCRIPTION
**Pull request by ontley**
#4393 

Includes willRenameFiles and didRenameFiles requests to the language server.
Honestly not sure if didRenameFiles is even needed here

Should file changes suggested by the language server be automatically saved if the file didn't have any other changes?

### CODE REVIEW

**Comment by the-mikedavis** [comment]
There is already https://github.com/helix-editor/helix/pull/4514 but it doesn't cover the LSP renaming

**Comment by cor** [comment]
Nice one! Really happy that it includes LSP rename support 👍🏻

**Comment by ontley** [comment]
I added the rnm alias for the command, referring to [#4514 (comment)](https://github.com/helix-editor/helix/pull/4514#issuecomment-1396122027)

**Comment by ontley** [comment]
I wonder if it would be good to add a feature to automatically save suggested changes from the lsp if a file didn't have changes before, saves the trouble of opening every file and saving manually, or perhaps a status message saying which files changed?

**Comment by devgioele** [comment]
Looking forward to this!

**Comment by edrex** [comment]
I need to test this, but do suggested changes cause non-open files to be opened?

> I wonder if it would be good to add a feature to automatically save suggested changes from the lsp if a file didn't have changes before, saves the trouble of opening every file and saving manually

IMO it's not needed, as `:wa` is easy enough, and there's already `editor.auto-save`. I get wanting transactionality, since the rename is immediate, but IDT it warrants a special case.

**Comment by ontley** [comment]
> I need to test this, but do suggested changes cause non-open files to be opened?
> 
> > I wonder if it would be good to add a feature to automatically save suggested changes from the lsp if a file didn't have changes before, saves the trouble of opening every file and saving manually
> 
> IMO it's not needed, as `:wa` is easy enough, and there's already `editor.auto-save`. I get wanting transactionality, since the rename is immediate, but IDT it warrants a special case.

**Comment by ontley** [comment]
> I need to test this, but do suggested changes cause non-open files to be opened?
> 
> > I wonder if it would be good to add a feature to automatically save suggested changes from the lsp if a file didn't have changes before, saves the trouble of opening every file and saving manually
> 
> IMO it's not needed, as `:wa` is easy enough, and there's already `editor.auto-save`. I get wanting transactionality, since the rename is immediate, but IDT it warrants a special case.

Ah I didn't even know about `:wa`, I guess it's not needed then

**Comment by edrex** [comment]
> Ah I didn't even know about :wa
It's a vimism. There's also `:wqa`.

**Comment by ontley** [comment]
Should the lsp errors be displayed in the status bar, along with the message that the file was successfully renamed?

**Comment by ontley** [comment]
> Why it is not merged already?

Waiting for reviews

**Comment by vikaskaliramna07** [comment]
> > Why it is not merged already?
> 
> Waiting for reviews

Can anyone review or contributors only?

**Comment by ontley** [comment]
> > > Why it is not merged already?
> > 
> > 
> > Waiting for reviews
> 
> Can anyone review or contributors only?

Only "reviewers with write access" and I can't even request reviews

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [inline helix-term/src/commands/typed.rs:2185]:
```suggestion
        .path()
```

using a relative path makes no sense here

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [inline helix-term/src/commands/typed.rs:2200]: i think that this should probably happen before the rename. If the fails later we can roll this back pretty easily (ideally with an unchecked method) while rolling back IO is pretty hard

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [inline helix-term/src/commands/typed.rs:2213]: we already show an error message herer you don't need to display it. The returned result is just forcases where we need to communicate to the LSP that the edit failed (not here)

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [inline helix-term/src/commands/typed.rs:2216]:
```suggestion
                    Err(err) => log::error!("willRename request failed: {err}"),
```

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [inline helix-term/src/commands/typed.rs:2202]: the wii rename request must be sent (and processed) *before* the rename happens so this needs to be moved to the start of the function:

> The will rename files request is sent from the client to the server before files are actually renamed as long as the rename is triggered from within the client either by a user action or by applying a workspace edit. The request can return a WorkspaceEdit which will be applied to workspace before the files are renamed. 

At the end of the function you must send didRename:

> The did rename files notification is sent from the client to the server when files were renamed from within the client.

you also need to support mullti LSP here. Probably ok to just send wiiRename to the first server that supports it and didRename to all of them

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs:2190]: note that this means that the destination directory might not exist so we need `rename!` too

File: helix-lsp/src/client.rs
Comment (pascalkuthe) [inline helix-lsp/src/client.rs:615]: this must check the server capabiliy `workspace.fileOperations.didRename`

File: helix-lsp/src/client.rs
Comment (pascalkuthe) [inline helix-lsp/src/client.rs:604]: you need to check `workspace.fileOperations.willRename` herer (or even better at the callsite to support multiple LSPs)

**Review by pascalkuthe — CHANGES_REQUESTED** [review]

**Comment by pascalkuthe** [comment]
> > > Why it is not merged already?
> > 
> > 
> > Waiting for reviews
> 
> Can anyone review or contributors only?

you are probably new to opensource. A PR needs to be reviewed by a miantainer before its merged. Only maintainers can review PRs. We can't just trust random people with code review. There are hundreds of issues and PRs for us to get through so it can take a while.

**Comment by vikaskaliramna07** [comment]
> > > > Why it is not merged already?
> > > 
> > > 
> > > Waiting for reviews
> > 
> > 
> > Can anyone review or contributors only?
> 
> you are probably new to opensource. A PR needs to be reviewed by a miantainer before its merged. Only maintainers can review PRs. We can't just trust random people with code review. There are hundreds of issues and PRs for us to get through so it can take a while.

yeah, I am new. Thanks for providing information about reviews!

File: helix-term/src/commands/typed.rs
Comment (ontley) [reply helix-term/src/commands/typed.rs:2817]: If we do set the alias to move, we're probably better of with calling the command itself `move` or `move_file`

File: helix-term/src/commands/typed.rs
Comment (ontley) [reply helix-term/src/commands/typed.rs:2190]: This is a sort of weird one, if we change the command name to `move` as previously said, probably a good ide, thoguh `mv` in bash takes absolute paths, not ones relative to the file but I think it's work

On the other hand I don't think a `file` renaming function should be able to move it as most file managers just let you rename it

Imo, `move` is probably better

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs:2817]: we have other commands that have totally different aliases too so I think leaving it like that is fine too

File: helix-term/src/commands/typed.rs
Comment (pascalkuthe) [reply helix-term/src/commands/typed.rs:2190]: I guess it might make mose sense to just call the entire command move and make the destination a path relative to CWD. If somebody wants to just change the filename they can use `%filname/../<new_name>` in the future.

I think that generally move is a better idea than rename because it can do more (and the LS can handle full moves  for example RA can move a module to a totally different module

File: helix-term/src/commands/typed.rs
Comment (wetheredge) [inline helix-term/src/commands/typed.rs:2186]:
```suggestion
        .ok_or_else(|| anyhow!("Scratch buffers can not be renamed, use :write"))?;
```

**Pull request closed without merging**
